core/remotes/docker/fetcher.go

```diff
@@ -263,74 +439,120 @@ func (r dockerFetcher) FetchByDigest(ctx context.Context, dgst digest.Digest, op
 }
 
 func (r dockerFetcher) open(ctx context.Context, req *request, mediatype string, offset int64) (_ io.ReadCloser, retErr error) {
-	if mediatype == "" {
-		req.header.Set("Accept", "*/*")
-	} else {
-		req.header.Set("Accept", strings.Join([]string{mediatype, `*/*`}, ", "))
-	}
-	req.header.Set("Accept-Encoding", "zstd;q=1.0, gzip;q=0.8, deflate;q=0.5")
+	parallelism, chunkSize := r.config.Parallelism(), int64(r.config.ConcurrentDownloadChunkSize)
 
-	if offset > 0 {
-		// Note: "Accept-Ranges: bytes" cannot be trusted as some endpoints
-		// will return the header without supporting the range. The content
-		// range must always be checked.
-		req.header.Set("Range", fmt.Sprintf("bytes=%d-", offset))
-	}
+	req.setMediaType(mediatype)
+	req.header.Set("Accept-Encoding", "zstd;q=1.0, gzip;q=0.8, deflate;q=0.5")
+	req.setOffset(offset)
```

Comment:
still need some time to digest the changes in this function.

I wonder if we can incapsulate changes into a separate function (that has multi part fetch support), and based on if we need multi-part fetch, we all differeent functions.

Reply:
Understood, and as you prefer ! 😊 My thinking here is that when parallel layer fetch is off, this is the exact same function, so this should be okay. LMK if anything

Reply:
Hey @djdongjin did you have enough time to digest these changes ? Are there any things I could do to make merging easier ?

Originally I was looking for sort of a plugin system that would let me create my own 'puller', it would have been nice to be able to switch back and forth.

Reply:
Hi @azr I will take a look today or early next week.

> Originally I was looking for sort of a plugin system that would let me create my own 'puller'

We actually have something similar :) instead of a plugin it's a separate systemd service. So the client will call the separate `puller` first which will download and put the blobs under containerd content store. So when the client later calls containerd to pull the image it's already there.

But I think it's great to add this to containerd itself and make the parallelism configurable.

Reply:
Hey @djdongjin ! Coming back to this thread as it's been a bit more than a week. Do you have any new? Thanks !

Reply:
sorry was quite busy these 2 weeks. I will try to find sometime this week!